api-server/src/server/boot/user.js

```diff
@@ -468,6 +471,93 @@ function postResetProgress(req, res, next) {
   );
 }
 
+function createPostResetProjectProgress(app) {
+  return async function postResetProjectProgress(req, res, next) {
+    const { User } = app.models;
+    const userId = req.user.id;
+    const { blocks } = req.body;
+
+    let toResetChallengeIds = [];
+    let undefinedBlocks = [];
+
+    let isSendBlocks = blocks && blocks.length > 0;
+    let isHasValidBlockToReset = false;
+    let isResetSuccessful = false;
+
+    if (isSendBlocks) {
+      for (const dashedBlockName of blocks) {
+        const blockChallengeIds = getChallengeIdsForBlock(dashedBlockName);
+        if (blockChallengeIds != undefined) {
+          toResetChallengeIds.push(...blockChallengeIds);
+          isHasValidBlockToReset = true;
+        } else {
+          undefinedBlocks.push(dashedBlockName);
+        }
+      }
+
+      if (isHasValidBlockToReset) {
+        try {
+          const result = await User.updateAll(
+            {
+              id: userId,
+              'completedChallenges.id': { in: toResetChallengeIds }
+            },
+            {
+              $pull: {
+                completedChallenges: { id: { in: toResetChallengeIds } }
+              }
+            }
+          );
+          isResetSuccessful = result.count > 0;
+        } catch (err) {
+          return next(err);
+        }
+      }
+    }
+
+    const resetMessage = buildResetMessage(
+      isSendBlocks,
+      isHasValidBlockToReset,
+      isResetSuccessful,
+      undefinedBlocks
+    );
+    return res.status(200).json({ message: resetMessage });
+  };
+
+  function buildResetMessage(
+    isSendBlocks,
+    isHasValidBlockToReset,
+    isResetSuccessful,
+    undefinedBlocks
+  ) {
+    if (!isSendBlocks) {
+      return `No blocks were provided for the reset operation.`;
+    }
+
+    let messageParts = [];
+
+    if (isResetSuccessful) {
+      messageParts.push(
+        `Challenges for the provided blocks have been successfully reset.`
+      );
+    } else if (isHasValidBlockToReset) {
+      messageParts.push(
+        `No challenges were reset. The user might not have completed any challenges for the requested blocks.`
+      );
+    }
+
+    if (undefinedBlocks.length > 0) {
+      messageParts.push(
+        `No challenges found for the following blocks: ${undefinedBlocks.join(', ')}`
+      );
+    } else if (!isHasValidBlockToReset && !isResetSuccessful) {
+      return `No valid challenges were found to reset for the requested blocks.`;
+    }
+
+    return messageParts.join(' ');
+  }
```

Comment:
The `message` should be something like `"flash.<message-short-name>"` (as seen in other routes), and the message short name should be added to the `client/i18n/locales/english/translations.json` file.

Also, these messages should be something the Camper (user) should read - not something mentioning "user"

Reply:
Thanks for clarifying, @ShaunSHamilton!

To be honest, I hadn't envisioned these messages as directly facing the users, especially since I referenced the `postResetProgress` that doesn't actually respond with any specific message. However, I really like the idea of incorporating user-friendly messages that can be utilized effectively.

I admit I'm not entirely familiar with how flash messages are integrated, but I'm eager to dive in and get up to speed. I'll start exploring how to work with flash messages and ensure our messages are aligned with what our Campers should read, following the `"flash.<message-short-name>"` pattern and adding the short names to the `client/i18n/locales/english/translations.json` file as you've suggested.

Thanks again for pointing this out! I'll get started on this right away!